keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt

```diff
@@ -344,4 +379,37 @@ class ApplicationService(
 
   private fun ConstraintState.toConstraintSummary() =
     StatefulConstraintSummary(type, status, createdAt, judgedBy, judgedAt, comment, attributes)
+
+  private fun getArtifactInstance(
+    artifact: DeliveryArtifact,
+    version: String
+  ): PublishedArtifact? {
+    val releaseStatus = repository.getReleaseStatus(artifact, version)
+    return repository.getArtifactInstance(artifact.name, artifact.type, version, releaseStatus)
+  }
+
+  // Generating a SCM diff link between source and target versions (the order does matter!)
+  private fun generateDiffLink(baseUrl: String?, newerGitMetadata: GitMetadata?, olderGitMetadata: GitMetadata?): String? {
+    return if (baseUrl != null && newerGitMetadata != null && olderGitMetadata != null) {
+          "$baseUrl/projects/${newerGitMetadata.project}/repos/${newerGitMetadata.repo?.name}/compare/commits?" +
+            "targetBranch=${olderGitMetadata.commitInfo?.sha}&sourceBranch=${newerGitMetadata.commitInfo?.sha}"
+    } else {
+      null
+    }
+  }
+
+  // Calling igor to fetch all base urls by SCM type, and returning the right one based on current commit link
+  private fun getScmBaseLink(commitLink: String): String? {
+    val scmInfo = runBlocking {
+      scmInfo.getSCMInfo()
+    }
+    //TODO[gyardeni]: replace this parsing when rocket will add scm type to gitMetadata
+    val currentScm = commitLink.substring(commitLink.indexOf("/") +2, commitLink.indexOf("."))
+    when(currentScm) {
+      "stash" ->
```

Comment:
Maybe a bit less brittle this way?
```suggestion
    when {
      "stash" in commitLink ->
```

Reply:
oh totally!